File: Projetos/ACBrLib/Demos/Java/BAL/Imports/ACBrLibBAL/src/com/acbr/bal/ACBrBAL.java
package com.acbr.bal;

import com.acbr.ACBrLibBase;
import com.acbr.ACBrSessao;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;
import com.sun.jna.ptr.DoubleByReference;
import com.sun.jna.ptr.FloatByReference;
import com.sun.jna.ptr.IntByReference;
import java.io.File;
import java.nio.ByteBuffer;
import java.nio.file.Paths;

public final class ACBrBAL extends ACBrLibBase {
    private interface ACBrBALLib extends Library {
        static String JNA_LIBRARY_NAME = LibraryLoader.getLibraryName();
        public final static ACBrBALLib INSTANCE = LibraryLoader.getInstance();
        
        class LibraryLoader {
            private static String library = "";
            private static ACBrBALLib instance = null;
                        
            private static String getLibraryName() {
                if ( library.isEmpty() ) {
                    library = Platform.is64Bit() ? "ACBrBAL64" : "ACBrBAL32";
                }  
                return library;
            }
            
            public static ACBrBALLib getInstance() {
                if ( instance == null ) {
                    instance = (ACBrBALLib) Native.synchronizedLibrary(
                            (Library) Native.loadLibrary(JNA_LIBRARY_NAME, ACBrBALLib.class));
                }
                
                return instance;
            }                             
        }
    
        int BAL_Inicializar(String eArqConfig, String eChaveCrypt);

        int BAL_Finalizar();

        int BAL_Nome(ByteBuffer buffer, IntByReference bufferSize);

        int BAL_Versao(ByteBuffer buffer, IntByReference bufferSize);

        int BAL_UltimoRetorno(ByteBuffer buffer, IntByReference bufferSize);

        int BAL_ConfigLer(String eArqConfig);

        int BAL_ConfigGravar(String eArqConfig);

        int BAL_ConfigLerValor(String eSessao, String eChave, ByteBuffer buffer, IntByReference bufferSize);

        int BAL_ConfigGravarValor(String eSessao, String eChave, String valor);

        int BAL_Ativar();

        int BAL_Desativar();

        int BAL_LePeso(int MillisecTimeOut, DoubleByReference Peso);

        int BAL_SolicitarPeso();

        int BAL_UltimoPesoLido(DoubleByReference Peso);

        int BAL_InterpretarRespostaPeso(ByteBuffer eResposta, FloatByReference Peso);                   
    }

    public ACBrBAL() throws Exception{
        File iniFile = Paths.get(System.getProperty("user.dir"), "ACBrLib.ini").toFile();
        if ( !iniFile.exists() ) {
            iniFile.createNewFile();
        }
        
        int ret = ACBrBALLib.INSTANCE.BAL_Inicializar(toUTF8(iniFile.getAbsolutePath()), toUTF8(""));
        checkResult(ret);
    }
    
    public ACBrBAL(String eArqConfig, String eChaveCrypt) throws Exception{
        int ret = ACBrBALLib.INSTANCE.BAL_Inicializar(toUTF8(eArqConfig), toUTF8(eChaveCrypt));
        checkResult(ret);
    }
    
    @Override
    protected void finalize() throws Throwable {
        try {
            int ret = ACBrBALLib.INSTANCE.BAL_Finalizar();
            checkResult(ret);
        } catch (Exception e) {
            super.finalize();
        }
    }
    
    public String nome() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_Nome(buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public String versao() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_Versao(buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public void configLer() throws Exception {
        configLer("");
    }

    public void configLer(String eArqConfig) throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_ConfigLer(toUTF8(eArqConfig));
        checkResult(ret);
    }

    public void configGravar() throws Exception {
        configGravar("");
    }

    public void configGravar(String eArqConfig) throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_ConfigGravar(toUTF8(eArqConfig));
        checkResult(ret);
    }

    public String configLerValor(ACBrSessao eSessao, String eChave) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_ConfigLerValor(toUTF8(eSessao.name()), toUTF8(eChave), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public void configGravarValor(ACBrSessao eSessao, String eChave, Object value) throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_ConfigGravarValor(toUTF8(eSessao.name()), toUTF8(eChave), toUTF8(value.toString()));
        checkResult(ret);
    }
    
    public void ativar() throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_Ativar();
        checkResult(ret);
    }
    
    public void desativar() throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_Desativar();
        checkResult(ret);
    }
    
    public double lePeso(int MillisecTimeOut) throws Exception {
        DoubleByReference peso = new DoubleByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_LePeso(MillisecTimeOut, peso);
        checkResult(ret);       
        
        System.out.println(peso.getValue());
        
        return peso.getValue();
    }

    public void SolicitarPeso() throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_SolicitarPeso();
        checkResult(ret);
    }

    public double ultimoPesoLido() throws Exception {
        DoubleByReference peso = new DoubleByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_UltimoPesoLido(peso);
        checkResult(ret);
        
        return peso.getValue();
    }

    public float interpretarRespostaPeso(ByteBuffer eResposta, double var) throws Exception {
        FloatByReference peso = new FloatByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_InterpretarRespostaPeso(eResposta, peso);
        checkResult(ret);
        
        return peso.getValue();
    } 
    
    @Override
    protected void UltimoRetorno(ByteBuffer buffer, IntByReference bufferLen) {
        ACBrBALLib.INSTANCE.BAL_UltimoRetorno(buffer, bufferLen);
    }
       
}

File: Projetos/ACBrLib/Demos/C#/BAL/Imports/Dinamico/ACBrLib.BAL/ACBrBAL.cs
using System.Runtime.InteropServices;
using System.Text;
using ACBrLib.Core;

namespace ACBrLib.BAL
{
    public sealed class ACBrBAL : ACBrLibHandle
    {
        #region InnerTypes

        private class Delegates
        {
            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_Inicializar(string eArqConfig, string eChaveCrypt);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_Finalizar();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_Nome(StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_Versao(StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_UltimoRetorno(StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_ConfigLer(string eArqConfig);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_ConfigGravar(string eArqConfig);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_ConfigLerValor(string eSessao, string eChave, StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_ConfigGravarValor(string eSessao, string eChave, string valor);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_Ativar();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_Desativar();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_LePeso(int MillisecTimeOut, ref double peso);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_SolicitarPeso();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_UltimoPesoLido(ref double peso);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int BAL_InterpretarRespostaPeso(string resposta, ref double peso);
        }

        #endregion InnerTypes

        #region Constructors

        public ACBrBAL(string eArqConfig = "", string eChaveCrypt = "") :
            base("ACBrBAL64.dll", "ACBrBAL32.dll")
        {
            var inicializar = GetMethod<Delegates.BAL_Inicializar>();
            var ret = ExecuteMethod(() => inicializar(ToUTF8(eArqConfig), ToUTF8(eChaveCrypt)));

            CheckResult(ret);
        }

        #endregion Constructors

        #region Properties

        public string Nome
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<Delegates.BAL_Nome>();
                var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public string Versao
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<Delegates.BAL_Versao>();
                var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        #endregion Properties

        #region Methods

        #region Ini

        public void ConfigGravar(string eArqConfig = "ACBrLib.ini")
        {
            var gravarIni = GetMethod<Delegates.BAL_ConfigGravar>();
            var ret = ExecuteMethod(() => gravarIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public void ConfigLer(string eArqConfig = "ACBrLib.ini")
        {
            var lerIni = GetMethod<Delegates.BAL_ConfigLer>();
            var ret = ExecuteMethod(() => lerIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public T ConfigLerValor<T>(ACBrSessao eSessao, string eChave)
        {
            var method = GetMethod<Delegates.BAL_ConfigLerValor>();

            var bufferLen = BUFFER_LEN;
            var pValue = new StringBuilder(bufferLen);
            var ret = ExecuteMethod(() => method(ToUTF8(eSessao.ToString()), ToUTF8(eChave), pValue, ref bufferLen));
            CheckResult(ret);

            var value = ProcessResult(pValue, bufferLen);
            return ConvertValue<T>(value);
        }

        public void ConfigGravarValor(ACBrSessao eSessao, string eChave, object value)
        {
            if (value == null) return;

            var method = GetMethod<Delegates.BAL_ConfigGravarValor>();
            var propValue = ConvertValue(value);

            var ret = ExecuteMethod(() => method(ToUTF8(eSessao.ToString()), ToUTF8(eChave), ToUTF8(propValue)));
            CheckResult(ret);
        }

        #endregion Ini

        public void Ativar()
        {
            var method = GetMethod<Delegates.BAL_Ativar>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public void Desativar()
        {
            var method = GetMethod<Delegates.BAL_Desativar>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public decimal LePeso(int MillisecTimeOut = 1000)
        {
            var peso = 0D;
            var method = GetMethod<Delegates.BAL_LePeso>();
            var ret = ExecuteMethod(() => method(MillisecTimeOut, ref peso));

            CheckResult(ret);

            return (decimal)peso;
        }

        public void SolicitarPeso()
        {
            var method = GetMethod<Delegates.BAL_SolicitarPeso>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public decimal UltimoPesoLido()
        {
            var peso = 0D;
            var method = GetMethod<Delegates.BAL_UltimoPesoLido>();
            var ret = ExecuteMethod(() => method(ref peso));

            CheckResult(ret);

            return (decimal)peso;
        }

        public decimal InterpretarRespostaPeso(string resposta)
        {
            var peso = 0D;
            var method = GetMethod<Delegates.BAL_InterpretarRespostaPeso>();
            var ret = ExecuteMethod(() => method(ToUTF8(resposta), ref peso));

            CheckResult(ret);

            return (decimal)peso;
        }

        #region Private Methods

        protected override void InitializeMethods()
        {
            AddMethod<Delegates.BAL_Inicializar>("BAL_Inicializar");
            AddMethod<Delegates.BAL_Finalizar>("BAL_Finalizar");
            AddMethod<Delegates.BAL_Nome>("BAL_Nome");
            AddMethod<Delegates.BAL_Versao>("BAL_Versao");
            AddMethod<Delegates.BAL_UltimoRetorno>("BAL_UltimoRetorno");
            AddMethod<Delegates.BAL_ConfigLer>("BAL_ConfigLer");
            AddMethod<Delegates.BAL_ConfigGravar>("BAL_ConfigGravar");
            AddMethod<Delegates.BAL_ConfigLerValor>("BAL_ConfigLerValor");
            AddMethod<Delegates.BAL_ConfigGravarValor>("BAL_ConfigGravarValor");
            AddMethod<Delegates.BAL_Ativar>("BAL_Ativar");
            AddMethod<Delegates.BAL_Desativar>("BAL_Desativar");
            AddMethod<Delegates.BAL_LePeso>("BAL_LePeso");
            AddMethod<Delegates.BAL_SolicitarPeso>("BAL_SolicitarPeso");
            AddMethod<Delegates.BAL_UltimoPesoLido>("BAL_UltimoPesoLido");
            AddMethod<Delegates.BAL_InterpretarRespostaPeso>("BAL_InterpretarRespostaPeso");
        }

        protected override void FinalizeLib()
        {
            var finalizar = GetMethod<Delegates.BAL_Finalizar>();
            var ret = ExecuteMethod(() => finalizar());
            CheckResult(ret);
        }

        protected override string GetUltimoRetorno(int iniBufferLen = 0)
        {
            var bufferLen = iniBufferLen < 1 ? BUFFER_LEN : iniBufferLen;
            var buffer = new StringBuilder(bufferLen);
            var ultimoRetorno = GetMethod<Delegates.BAL_UltimoRetorno>();

            if (iniBufferLen < 1)
            {
                ExecuteMethod(() => ultimoRetorno(buffer, ref bufferLen));
                if (bufferLen <= BUFFER_LEN) return FromUTF8(buffer);

                buffer.Capacity = bufferLen;
            }

            ExecuteMethod(() => ultimoRetorno(buffer, ref bufferLen));
            return FromUTF8(buffer);
        }

        #endregion Private Methods

        #endregion Methods
    }
}